Bearded.Utilities/Noise/PerlinNoise.cs

```diff
@@ -0,0 +1,113 @@
+{
+    public static class PerlinNoise
+    {
+        // Use uniformly distributed vectors to force better gradient distribution.
+        private const int numVectorSamples = 12;
+        private static readonly Vector2d[] vectorSamples = vectorSamples = Enumerable.Range(0, numVectorSamples)
+            .Select(i => Direction2.FromRadians(MathConstants.TwoPi * i / numVectorSamples).Vector)
+            .Select(vector => new Vector2d(vector.X, vector.Y))
+            .ToArray();
+
+        /// <summary>
+        /// Generates a procedural texture using Perlin noise.
+        /// </summary>
+        /// <remarks>
+        /// Perlin noise is generated by first generating a grid of gradient vectors. The texture is then filled in by
+        /// for each point identifying the four surrounding gradient vectors, calculating the dot product between each
+        /// of the gradient vectors and corresponding offset vector, and interpolating between them.
+        /// </remarks>
+        /// <param name="numCellsX">
+        /// The number of cells in which the texture is divided along the horizontal axis (see remarks).
+        /// </param>
+        /// <param name="numCellsY">
+        /// The number of cells in which the texture is divided along the vertical axis (see remarks).
+        /// </param>
+        /// <param name="seed"></param>
+        /// <returns></returns>
+        /// <exception cref="ArgumentOutOfRangeException"></exception>
+        public static IProceduralTexture Generate(int numCellsX, int numCellsY, int? seed)
+        {
+            if (numCellsX <= 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(numCellsX));
+            }
+            if (numCellsY <= 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(numCellsY));
+            }
+
+            var gradientArray = NoiseUtils.GenerateRandomArray(
+                // We generate the corners, but to make the noise map wrap along both dimensions, we reuse the same
+                // values as the left and top boundaries, so we don't have to generate the right and bottom boundaries.
+                numCellsX,
+                numCellsY,
+                r => vectorSamples[r.Next(vectorSamples.Length)],
+                seed);
+            return new PerlinProceduralTexture(gradientArray);
+        }
+
+        private sealed class PerlinProceduralTexture : IProceduralTexture
+        {
+            private readonly Vector2d[,] gradientArray;
+            private readonly int width;
+            private readonly int height;
+
+            public PerlinProceduralTexture(Vector2d[,] gradientArray)
+            {
+                this.gradientArray = gradientArray;
+                width = gradientArray.GetLength(0);
+                height = gradientArray.GetLength(1);
+            }
+
+            public double ValueAt(double x, double y)
+            {
+                if (x < 0 || x >= 1)
```

Comment:
Yes, I would like to keep these. Sure, the exceptions will get thrown eventually, but with different numbers and in a different context. I prefer early failure in this case, because it is more obvious what the wrong in parameters are.